Source/Core/Common/VR/DolphinVR.cpp

```diff
@@ -0,0 +1,239 @@
+
+  // Left controller to GCPad
+  SetControlState(0, GCPAD_X_BUTTON, leftController & xrButton_Trigger ? 1 : 0);
+  SetControlState(0, GCPAD_Y_BUTTON, leftController & xrButton_GripTrigger ? 1 : 0);
+  SetControlState(0, GCPAD_L_ANALOG, leftController & xrButton_X ? 1 : 0);
+  SetControlState(0, GCPAD_R_ANALOG, leftController & xrButton_Y ? 1 : 0);
+  SetControlState(0, GCPAD_L_DIGITAL, leftController & xrButton_X ? 1 : 0);
+  SetControlState(0, GCPAD_R_DIGITAL, leftController & xrButton_Y ? 1 : 0);
+
+  // Left controller stick tn GCPad
+  if (rightController & xrButton_B)
+  {
+    SetControlState(0, GCPAD_C_STICK_X, IN_VRGetJoystickState(0).x);
+    SetControlState(0, GCPAD_C_STICK_Y, IN_VRGetJoystickState(0).y);
+  }
+  else
+  {
+    SetControlState(0, GCPAD_MAIN_STICK_X, IN_VRGetJoystickState(0).x);
+    SetControlState(0, GCPAD_MAIN_STICK_Y, IN_VRGetJoystickState(0).y);
+  }
+
+  // Right controller to GCPad
+  SetControlState(0, GCPAD_A_BUTTON, rightController & xrButton_Trigger ? 1 : 0);
+  SetControlState(0, GCPAD_B_BUTTON, rightController & xrButton_GripTrigger ? 1 : 0);
+  SetControlState(0, GCPAD_Z_BUTTON, rightController & xrButton_A ? 1 : 0);
+  SetControlState(0, GCPAD_DPAD_UP, rightController & xrButton_Up ? 1 : 0);
+  SetControlState(0, GCPAD_DPAD_DOWN, rightController & xrButton_Down ? 1 : 0);
+  SetControlState(0, GCPAD_DPAD_LEFT, rightController & xrButton_Left ? 1 : 0);
+  SetControlState(0, GCPAD_DPAD_RIGHT, rightController & xrButton_Right ? 1 : 0);
+  SetControlState(0, GCPAD_START_BUTTON, rightController & xrButton_RThumb ? 1 : 0);
+
+  // Left controller to WiiMote + Nunchuk
+  SetControlState(0, WIIMOTE_PLUS_BUTTON, leftController & xrButton_X ? 1 : 0);
+  SetControlState(0, WIIMOTE_MINUS_BUTTON, leftController & xrButton_Y ? 1 : 0);
+  SetControlState(0, WIIMOTE_HOME_BUTTON, leftController & xrButton_Back ? 1 : 0);
+  SetControlState(0, NUNCHUK_C_BUTTON, leftController & xrButton_Trigger ? 1 : 0);
+  SetControlState(0, NUNCHUK_Z_BUTTON, leftController & xrButton_GripTrigger ? 1 : 0);
+  SetControlState(0, NUNCHUK_STICK_X, IN_VRGetJoystickState(0).x);
+  SetControlState(0, NUNCHUK_STICK_Y, IN_VRGetJoystickState(0).y);
+
+  // Right controller to WiiMote
+  SetControlState(0, WIIMOTE_A_BUTTON, rightController & xrButton_Trigger ? 1 : 0);
+  SetControlState(0, WIIMOTE_B_BUTTON, rightController & xrButton_GripTrigger ? 1 : 0);
+  SetControlState(0, WIIMOTE_ONE_BUTTON, rightController & xrButton_A ? 1 : 0);
+  SetControlState(0, WIIMOTE_TWO_BUTTON, rightController & xrButton_B ? 1 : 0);
+  SetControlState(0, WIIMOTE_DPAD_UP, rightController & xrButton_Up ? 1 : 0);
+  SetControlState(0, WIIMOTE_DPAD_DOWN, rightController & xrButton_Down ? 1 : 0);
+  SetControlState(0, WIIMOTE_DPAD_LEFT, rightController & xrButton_Left ? 1 : 0);
+  SetControlState(0, WIIMOTE_DPAD_RIGHT, rightController & xrButton_Right ? 1 : 0);
+  SetControlState(0, WIIMOTE_IR_X, x);
+  SetControlState(0, WIIMOTE_IR_Y, y);
```

Comment:
Handling controller mappings in this way is probably OK for early testing, but I would not be comfortable with merging this. Without the ability to remap controls, playing games that require a Wii Remote can be very troublesome.

Reply:
The controller mapping is quite tricky. In the Android APK (debug or release) which I use as a game launcher, I have no access to OpenXR API, so I cannot detect user's input.

In the future I plan to create a new UI which I can render inside OpenXR, then there will be possibility to remap controllers. New UI will also be a step to get rid of the need of two APKs.

The VR support is a big long term development, I do not want to pack too much into one PR.

Reply:
I suppose that's fair. Tricky problem...